runtime/ts/keymgmt/webcrypto.ts

```diff
@@ -0,0 +1,276 @@
+            true,
+            ["encrypt", "decrypt"]
+        ).then(key => new WebCryptoSessionKey(key));
+    }
+
+    rewrap(privKey: PrivateKey, pubKey: PublicKey): PromiseLike<WrappedKey> {
+        return this.unwrap(privKey).then(skey => skey.disposeToWrappedKeyUsing(pubKey));
+    }
+
+    export(): string {
+        return encode(this.wrappedKeyData.buffer as ArrayBuffer);
+    }
+
+    fingerprint(): PromiseLike<string> {
+        return Promise.resolve(encode(this.wrappedKeyData.buffer as ArrayBuffer));
+    }
+}
+
+class WebCryptoPrivateKey extends WebCryptoStorableKey<CryptoKey> implements PrivateKey {
+    constructor(key) {
+        super(key);
+    }
+
+    cryptoKey() {
+        return this.storableKey();
+    }
+
+    fingerprint(): PromiseLike<string> {
+        return Promise.resolve("");
+    }
+}
+
+class WebCryptoPublicKey extends WebCryptoStorableKey<CryptoKey> implements PublicKey {
+
+    constructor(key) {
+        super(key);
+    }
+
+    cryptoKey() {
+        return this.storableKey();
+    }
+
+    fingerprint(): PromiseLike<string> {
+        // TODO: fix this with a proper hash based fingerprint/thumbprint
+        return crypto.subtle.exportKey("jwk", this.cryptoKey()).then(key => JSON.stringify(key));
+    }
+}
+
+class WebCryptoSessionKey implements SessionKey {
+    decrypt(buffer: ArrayBuffer): PromiseLike<ArrayBuffer> {
+        return crypto.subtle.decrypt({
+            name: this.algorithm(),
+            iv: this.iv,
+        }, this.sessionKey, buffer);
+    }
+
+    encrypt(buffer: ArrayBuffer): PromiseLike<ArrayBuffer> {
+        return crypto.subtle.encrypt(
+            {
+                name: this.algorithm(),
+                iv: this.iv
+            }, this.sessionKey, buffer);
+    }
+
+    sessionKey: CryptoKey;
+    // Cached IV only for testing currently
+    iv: Uint8Array;
+
+    constructor(sessionKey: CryptoKey) {
+        this.sessionKey = sessionKey;
+        // hack for unit testing
+        this.iv = new Uint8Array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12]);
+    }
+
+    algorithm(): string {
+        return "AES-GCM";
+    }
+
+    disposeToWrappedKeyUsing(pkey: PublicKey): PromiseLike<WrappedKey> {
+        try {
+            const webPkey = pkey as WebCryptoPublicKey;
+
+            const rawWrappedKey = crypto.subtle.wrapKey("raw",
+                this.sessionKey,
+                (pkey as WebCryptoPublicKey).cryptoKey(),
+                {   //these are the wrapping key's algorithm options
+                    name: webPkey.algorithm(),
+                }
+            );
+            return rawWrappedKey.then(rawKey => new WebCryptoWrappedKey(new Uint8Array(rawKey), pkey));
+        } finally {
+            // Hopefully this frees the underlying key material
+            this.sessionKey = null;
+        }
+    }
+
+    isDisposed(): boolean {
+        return this.sessionKey != null;
+    }
+
+    fingerprint(): PromiseLike<string> {
+        return Promise.resolve("");
```

Comment:
Revise in some manner as discussed -- perhaps base class throws exception, subclasses that actually allow fingerprinting override.